spyder/api/mode.py

```diff
@@ -0,0 +1,134 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Copyright © <2013-2016> <Colin Duquesnoy and others, see pyqode/AUTHORS.rst>
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""
+This module contains the editor extension API.
+Adapted from https://github.com/pyQode/pyqode.core/blob/master/pyqode/core/api/mode.py
+"""
+import logging
+
+
+def _logger():
+    return logging.getLogger(__name__)
+
+
+class Mode(object):
+    """
+    Base class for editor extensions. An extension is a "thing" that can be
+    installed on an editor to add new behaviours or to modify its appearance.
+
+    At the current states Modes can't be added to an Editor but this class
+    is needed because Panels can be added to an editor.
+
+    A panel (model child class) is added to an editor by using the
+    PanelsManager:
+        - :meth:`spyder.widgets.sourcecode.CodeEditor.panels.append`
+
+    Subclasses may/should override the following methods:
+
+        - :meth:`spyder.api.Mode.on_install`
+        - :meth:`spyder.api.Mode.on_uninstall`
+        - :meth:`spyder.api.Mode.on_state_changed`
+
+    ..warning: The mode will be identified by its class name, this means that
+    **there cannot be two modes of the same type on the same editor instance!**
+    """
+
+    @property
+    def editor(self):
+        """
+        Returns a reference to the parent code editor widget.
+
+        **READ ONLY**
+
+        :rtype: spyder.widgets.sourcecode.CodeEditor
+        """
+        if self._editor is not None:
+            return self._editor
+        else:
+            return None
+
+    @property
+    def enabled(self):
+        """
+        Tells if the mode is enabled,
```

Comment:
Finish this line with a dot and add a blank line to separate it from the description